arbor/morph/region.cpp

```diff
@@ -0,0 +1,306 @@
+}
+
+mcable_list merge(const mcable_list& v) {
+    if (v.size()<2) return v;
+    std::vector<mcable> L;
+    L.reserve(v.size());
+    auto c = v.front();
+    auto it = v.begin()+1;
+    while (it!=v.end()) {
+        if (!is_disjoint(c, *it)) {
+            c = make_union(c, *it);
+        }
+        else {
+            L.push_back(c);
+            c = *it;
+        }
+        ++it;
+    }
+    L.push_back(c);
+    return L;
+}
+
+// Insert a zero-length region at the start of each child branch for every cable
+// that includes the end of a branch.
+// Insert a zero-length region at the end of the parent branch for each cable
+// that includes the start of the branch.
+mcable_list cover(mcable_list cables, const em_morphology& m) {
+    mcable_list L;
+    for (auto& c: cables) {
+        if (c.prox_pos==0) {
+            for (auto& x: m.cover(mlocation{c.branch, 0}, false)) {
+                L.push_back({x.branch, x.pos, x.pos});
+            }
+        }
+        if (c.dist_pos==1) {
+            for (auto& x: m.cover(mlocation{c.branch, 1}, false)) {
+                L.push_back({x.branch, x.pos, x.pos});
+            }
+        }
+    }
+    L.insert(L.end(), cables.begin(), cables.end());
+    util::sort(L);
+
+    return L;
+}
+
+mcable_list remove_cover(mcable_list cables, const em_morphology& m) {
+    // Find all zero-length cables at the end of cables, and convert to
+    // their canonical representation.
+    for (auto& c: cables) {
+        if (c.dist_pos==0 || c.prox_pos==1) {
+            auto cloc = m.canonicalize({c.branch, c.prox_pos});
+            c = {cloc.branch, cloc.pos, cloc.pos};
+        }
+    }
+    // Some new zero-length cables may be out of order, so sort
+    // the cables.
+    util::sort(cables);
+
+    // Remove multiple copies of zero-length cables if present.
+    return merge(cables);
+}
+
+//
+// Explicit cable section
+//
+
+struct cable_ {
+    mcable cable;
+};
+
+region cable(mcable c) {
+    if (!test_invariants(c)) {
+        throw morphology_error(util::pprintf("Invalid cable section {}", c));
+    }
+    return region(cable_{c});
+}
+
+region branch(msize_t bid) {
+    return cable({bid, 0, 1});
+}
+
+mcable_list thingify_(const cable_& reg, const em_morphology& em) {
+    auto& m = em.morph();
+
+    if (reg.cable.branch>=m.num_branches()) {
+        throw morphology_error(util::pprintf("Branch {} does not exist in morpology", reg.cable.branch));
+    }
+
+    return {reg.cable};
+}
+
+std::ostream& operator<<(std::ostream& o, const cable_& c) {
+    return o << c.cable;
+}
+
+//
+// region with all segments with the same numeric tag
+//
+struct tagged_ {
+    int tag;
+};
+
+region tagged(int id) {
+    return region(tagged_{id});
+}
+
+mcable_list thingify_(const tagged_& reg, const em_morphology& em) {
+    auto& m = em.morph();
+    size_t nb = m.num_branches();
+
+    std::vector<mcable> L;
+    L.reserve(nb);
```

Comment:
In many cases, the number of branches will be a significant over-estimate of the size of the cable list; is it worth reserving? Or, alternatively, is it worth doing a `shrink_to_fit` before returning it?

Reply:
For they typical soma case, it grossly underestimates. For dendrite not. On the balance of things, I can't tell what is worse: over allocating on regions like soma, or making many small allocations when building dendrites? I would say pick an approach, and see if it is an issue in production.

Reply:
Added a shrink to fit.